_posts/2020-06-01-spring-boot-hazelcast.md

```diff
@@ -0,0 +1,232 @@
+---
+title: Distributed Cache with Hazelcast and Spring.
+categories: [spring-boot]
+date: 2020-05-26 05:00:00 +1100
+modified: 2020-05-26 05:00:00 +1100
+author: artur
+excerpt: "Cache is needed for fast access to the data. Some applications need a distributed cache. This article shows how to set up a distributed cache cluster with Hazelcast and Spring."
+image:
+  auto: 0070-hazelcast
+---
+
+In some applications, it is needed to offload the database or avoid multiple calculations of cost-intensive operation.
+A cache is used for this goal. This article shows how to use [Hazelcast](https://hazelcast.org/) as Cache with Spring 
+in a distributed and scalable application.
+
+{% include github-project.html url="https://github.com/arkuksin/code-examples/tree/hazelcast/spring-boot/hazelcast" %}
+
+## Why Cache
+Normally, an application reads data from the storage, for example, a database. If we want to increase the performance
+of reading or writing data, we can improve the hardware and make it faster. But it costs money.
+If the data in the external storage don't change very fast, we can create copies of these data in smaller but
+much faster storage. These copies are store temporarily. Usually, the RAM is used for such kind of storage.
+It is called a cache.
+
+If the application wants to access data, it requests the data in the cache. We know, that the data in the cache are copies,
+and we cannot use them for a long time. The data in the primary storage can change. In this case, we would get a data inconsistency.
+That's why we need to define the validity time of the data in the cache. Also, we don't need the data in the cache, 
+that is not frequently required. These data allocate resources of the cache and are not used. In this case, we can configure the time,
+how long a data lives in the cache if it is not required. It is called **time-to-live** (TTL).
+
+In a big enterprise system, there can be a cluster of caches. We have to replicate and synchronize the data in this cluster
+between the caches. It is called **write-through concept**.
+
+## Hazelcast as Distributed Cache
+
+Let's say we have a Spring application, and we want to build in a cache in the application. But we also want to 
+be able to scale this application. It means, when we start for example three instances of the application, they have to
+share the cache to keep the data consistently. This problem can be solved by using a **distributed cache**.
+
+Hazelcast is a distributed in-memory object store and provides many features including TTL, work-through, and scalability.
+We can build a Hazelcast cluster by starting several `Hazelcast` nodes in a net. This node is called a member. 
```

Comment:
Product names should not be highlighted with mono-spaced font. I.e. it should be Hazelcast instead of `Hazelcast`. I changed that already throughout the article.